File: backend/src/main/java/com/headless/ecommerce/service/CategoryService.java
package com.headless.ecommerce.service;

import com.headless.ecommerce.dto.request.CategoryCreateRequest;
import com.headless.ecommerce.dto.response.CategoryResponse;

import java.util.List;

/**
 * Category service interface for category management.
 */
public interface CategoryService {

    /**
     * Gets the category tree.
     *
     * @return the list of root categories with children
     */
    List<CategoryResponse> getCategoryTree();

    /**
     * Creates a new category.
     *
     * @param request the category creation request
     * @return the created category response
     */
    CategoryResponse createCategory(CategoryCreateRequest request);

    /**
     * Updates an existing category.
     *
     * @param id the category ID
     * @param request the category update request
     * @return the updated category response
     */
    CategoryResponse updateCategory(Long id, CategoryCreateRequest request);

    /**
     * Deletes a category.
     *
     * @param id the category ID
     */
    void deleteCategory(Long id);
}

File: frontend/src/services/category-service.ts
import api from "@/lib/api";
import { API_PATHS } from "@/lib/constants";
import type { ApiResponse } from "@/types/api";
import type { Category, CategoryCreateRequest } from "@/types/category";

/**
 * Gets the category tree.
 */
export async function getCategories(): Promise<ApiResponse<Category[]>> {
  const response = await api.get<ApiResponse<Category[]>>(
    API_PATHS.CATEGORIES.LIST
  );
  return response.data;
}

/**
 * Creates a new category (admin).
 */
export async function createCategory(
  data: CategoryCreateRequest
): Promise<ApiResponse<Category>> {
  const response = await api.post<ApiResponse<Category>>(
    API_PATHS.ADMIN.CATEGORIES,
    data
  );
  return response.data;
}

/**
 * Updates a category (admin).
 */
export async function updateCategory(
  id: number,
  data: CategoryCreateRequest
): Promise<ApiResponse<Category>> {
  const response = await api.put<ApiResponse<Category>>(
    API_PATHS.ADMIN.CATEGORY_DETAIL(id),
    data
  );
  return response.data;
}

/**
 * Deletes a category (admin).
 */
export async function deleteCategory(id: number): Promise<ApiResponse<void>> {
  const response = await api.delete<ApiResponse<void>>(
    API_PATHS.ADMIN.CATEGORY_DETAIL(id)
  );
  return response.data;
}
